### ISSUE TITLE
Finalize optimizer API

### ISSUE THREAD

**Issue by mariaschuld** [PennyLaneAI/pennylane#47]
Being a bit behind from the vacation, I see that we still have the Optimizer.train() API. Can we change this into having a base Optimizer class from which we build AdamOptimizer, GradientDescentOptimizer, NelderMeadOptimizer...

The crucial function of those is "step" which updates the weights by one step. The Optimizer stores 
-the cost, 
-the gradient function, 
-the current weights, 
-some past gradients if necessary 
and has hyperparameters
-learning rate
-regulariser (although this could also be manually added to the cost)
-weight of momentum etc
-...

In particular, we should delete SGD from 0.1.
